tests/schema/tests.py

```diff
@@ -275,6 +275,18 @@ def test_inline_fk(self):
             if sql.startswith('ALTER TABLE') and 'ADD CONSTRAINT' in sql
         ])
 
+    @skipUnlessDBFeature('can_create_inline_fk')
+    def test_add_inline_fk_update_data(self):
+        with connection.schema_editor() as editor:
+            editor.create_model(Node)
+        # Add an inline foreign key and update data in the same transaction.
+        new_field = ForeignKey(Node, CASCADE, related_name='new_fk', null=True)
+        new_field.set_attributes_from_name('new_parent_fk')
+        parent = Node.objects.create()
+        with connection.schema_editor() as editor:
```

Comment:
I haven't checked out the code but I don't think this test is appropriate; [`deferred_sql` is only executed at `editor.__exit__`](https://github.com/django/django/blob/57a3d96ff57abffa8e7fbe3fd972a5ee39bd0fec/django/db/backends/base/schema.py#L112-L117) which means `SET CONSTRAINTS` will not be executed before the `UPDATE schema_node`.

Reply:
🤔 It raises _"django.db.utils.OperationalError: cannot CREATE INDEX ... because it has pending trigger events"_ without this patch.

Reply:
Error is raised on `CREATE INDEX`, so we have here three steps:
- add field with FK,
- update data,
- create index (which is deffered).

This patch sets constraint to `IMMEDIATE` before the 3rd step.

Reply:
That's weird, I cannot get the test to fail by commenting the above lines on PostgreSQL 10 using django-docker-box

```
> POSTGRES_VERSION=10 docker-compose run --rm postgres schema.tests.SchemaTests.test_add_inline_fk_update_data
Starting django-docker-box_postgres-db_1 ... done
Starting django-docker-box_memcached_1   ... done
Starting django-docker-box_memcached2_1  ... done
wait-for-it.sh: waiting 20 seconds for postgres-db:5432
wait-for-it.sh: postgres-db:5432 is available after 0 seconds
/usr/local/lib/python3.6/site-packages/IPython/paths.py:68: UserWarning: IPython parent '/home/test' is not a writable location, using a temp directory.
  " using a temp directory.".format(parent))
Testing against Django installed in '/tests/django/django' with up to 2 processes
Creating test database for alias 'default'...
System check identified no issues (0 silenced).

Running tests...
----------------------------------------------------------------------
.
----------------------------------------------------------------------
Ran 1 test in 0.074s

OK

Generating XML reports...
Destroying test database for alias 'default'...
```

In all case nothing would prevent a non-deferred index to be added in the same transaction using `editor.add_index(Node, Index(fields=['new_parent_fk'], name='new_parent_idx'))`.

Reply:
I'm getting:
```
POSTGRES_VERSION=10 docker-compose run --rm postgres schema.tests.SchemaTests.test_add_inline_fk_update_data --parallel=1       
Creating djangodockerbox_postgres-db_1 ... 
Starting djangodockerbox_memcached2_1 ... 
Creating djangodockerbox_postgres-db_1 ... done
wait-for-it.sh: waiting 20 seconds for postgres-db:5432
wait-for-it.sh: postgres-db:5432 is available after 1 seconds
Testing against Django installed in '/tests/django/django'
Creating test database for alias 'default'...
System check identified no issues (0 silenced).

Running tests...
----------------------------------------------------------------------
E
ERROR [0.000s]: test_add_inline_fk_update_data (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.ObjectInUse: cannot CREATE INDEX "schema_node" because it has pending trigger events
```

:face_with_head_bandage:

Reply:
@charettes IMO this test is appropriate, operations are in the same order as in a sample project attached to the ticket. i.e.:
```
ALTER TABLE "schema_node" ADD CONSTRAINT "schema_node_parent_id_5e700b70_fk_schema_node_node_id" FOREIGN KEY ("parent_id") REFERENCES "schema_node" ("node_id") DEFERRABLE INITIALLY DEFERRED ()
CREATE INDEX "schema_node_parent_id_5e700b70" ON "schema_node" ("parent_id") ()
INSERT INTO "schema_node" ("parent_id") VALUES (%s) RETURNING "schema_node"."node_id" (None,)
ALTER TABLE "schema_node" ADD COLUMN "new_parent_fk_id" integer NULL CONSTRAINT "schema_node_new_parent_fk_id_4d987561_fk_schema_node_node_id" REFERENCES "schema_node"("node_id") DEFERRABLE INITIALLY DEFERRED []
UPDATE schema_node SET new_parent_fk_id = %s; [1]
```
`editor.__exit__()`:
```
SET CONSTRAINTS "schema_node_new_parent_fk_id_4d987561_fk_schema_node_node_id" IMMEDIATE ()
CREATE INDEX "schema_node_new_parent_fk_id_4d987561" ON "schema_node" ("new_parent_fk_id") ()
```

I agree, that we should add a second test with `add_index()`.

Reply:
About new test with `add_index()`, I cannot use `add_index()` because it will always crash with `Node has no field named 'new_parent_fk'` (model is not aware of the new field).

I can use `alter_field()` and add `db_index=True`, but it's not affected by this issue, because it already sets constraints to `IMMEDIATE`:

https://github.com/django/django/blob/45bcc6feac68165eb3642d3c308c74b4828e679e/django/db/backends/postgresql/schema.py#L23-L25

That's why I did this manually.